File: bpf/cred_monitor.c
//go:build ignore

/*
 * Credential-change monitoring (CO-RE) via a kprobe on commit_creds.
 *
 * Every other monitor in Pahlevan watches a request: a file being opened, a
 * socket connecting, a program being executed, a capability being checked.
 * This one watches the outcome. commit_creds is the single kernel function
 * through which a task's credentials change - every setuid, every capability
 * grant, every successful privilege escalation, whatever route it took to get
 * there.
 *
 * That "whatever route" is the point. A syscall monitor sees setuid(0) and a
 * seccomp profile can forbid it, but neither sees a kernel exploit that
 * overwrites a cred struct and calls commit_creds directly - which is what the
 * overwhelming majority of local-root exploits do, because it is the shortest
 * path from an arbitrary write to a root shell. From here that exploit is not
 * subtle at all: a task's euid becomes 0, or its effective capability set
 * gains bits, and no execve is in progress to explain it.
 *
 * That last clause is the discriminator. A setuid binary legitimately gains
 * privilege, and it does so inside execve, where task->in_execve is set. A
 * process that gains root without an execve underway either called a setuid
 * syscall - which the syscall monitor already reports with its arguments - or
 * did something the kernel has no other name for.
 *
 * A kprobe cannot refuse the change: commit_creds returns void and is called
 * past the point of no return. It can, however, kill the process that made it,
 * before the new credentials are used for anything. That is the only
 * enforcement available at this site and it is worth having, because there is
 * no LSM hook here.
 *
 * Copyright 2025. Licensed under the Apache License, Version 2.0.
 */

#include "vmlinux.h"

/* bpf_tracing.h casts a kprobe's context to struct user_pt_regs on arm64, and
 * vmlinux.h is generated from whichever kernel built the image - in practice
 * an x86 one, which has no such type. Declaring it here is safe: it is a UAPI
 * structure, so its layout is a stable kernel ABI and cannot drift. Without
 * this the arm64 object does not compile at all, which is how the previous
 * attempt at reading syscall arguments was lost. */
#if defined(__TARGET_ARCH_arm64)
struct user_pt_regs {
	__u64 regs[31];
	__u64 sp;
	__u64 pc;
	__u64 pstate;
};
#endif

#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>

char LICENSE[] SEC("license") = "GPL";

#define MODE_OBSERVE 0
#define MODE_KILL    1

/* Flags on the event. */
#define CRED_GAINED_ROOT  0x01 /* euid became 0 and was not 0 before */
#define CRED_GAINED_CAPS  0x02 /* the effective capability set grew */
#define CRED_IN_EXECVE    0x04 /* an execve is underway; a setuid binary explains it */
#define CRED_KILLED       0x08 /* the task was sent SIGKILL */
#define CRED_LOST_ROOT    0x10 /* euid stopped being 0 - a daemon dropping privilege */

struct cred_event {
	__u64 cgroup_id;
	__u64 timestamp_ns;
	__u64 old_caps; /* cap_effective before */
	__u64 new_caps; /* cap_effective after */
	__u32 pid;      /* tgid */
	__u32 tid;
	__u32 old_uid;
	__u32 new_uid;
	__u32 old_euid;
	__u32 new_euid;
	__u32 flags;
	__u8  comm[16];
};

struct {
	__uint(type, BPF_MAP_TYPE_RINGBUF);
	__uint(max_entries, 1 << 20);
} cred_events SEC(".maps");

/* key 0: enabled, key 1: mode, key 2: kill even inside execve */
struct {
	__uint(type, BPF_MAP_TYPE_ARRAY);
	__uint(key_size, sizeof(__u32));
	__uint(value_size, sizeof(__u32));
	__uint(max_entries, 4);
} cred_config SEC(".maps");

/* Cgroups whose credential changes are governed. Empty means every cgroup is
 * observed and none is killed, which is the safe default for a monitor that
 * can only respond with a signal. */
struct {
	__uint(type, BPF_MAP_TYPE_HASH);
	__type(key, __u64);
	__type(value, __u32); /* mode for this cgroup */
	__uint(max_entries, 8192);
} cred_governed SEC(".maps");

static __always_inline __u32 cfg(__u32 key)
{
	__u32 *v = bpf_map_lookup_elem(&cred_config, &key);
	return v ? *v : 0;
}

/* The effective capability set, read in a way that survives the kernel 6.3
 * change of representation.
 *
 * Before 6.3 a kernel_cap_t was __u32 cap[2]; from 6.3 it is a single __u64
 * val. Both occupy the same eight bytes and, on the little-endian
 * architectures BPF targets here, cap[0] is the low half of val - so one
 * eight-byte read produces the same number either way, and no CO-RE branch on
 * the field name is needed. Only the offset of cap_effective within struct
 * cred is relocated, and that field exists under that name in both. */
static __always_inline __u64 effective_caps(const struct cred *c)
{
	__u64 caps = 0;
	bpf_core_read(&caps, sizeof(caps), &c->cap_effective);
	return caps;
}

/* Whether an execve is underway for the current task.
 *
 * in_execve is a bitfield, so it needs the bitfield-aware CO-RE read rather
 * than an ordinary one. If a kernel ever drops the field the read fails
 * closed - reporting "no execve underway", which makes a legitimate setuid
 * binary look like an escalation. That is the right direction to fail in: a
 * false report is investigated, a missed one is not. */
static __always_inline int in_execve(struct task_struct *task)
{
	if (!bpf_core_field_exists(task->in_execve))
		return 0;
	return BPF_CORE_READ_BITFIELD_PROBED(task, in_execve) != 0;
}

SEC("kprobe/commit_creds")
int BPF_KPROBE(handle_commit_creds, struct cred *new)
{
	if (!cfg(0))
		return 0;
	if (!new)
		return 0;

	struct task_struct *task = bpf_get_current_task_btf();
	if (!task)
		return 0;

	/* real_cred rather than cred: at this point cred is still the old one for
	 * the duration of the call, and real_cred is what the task was actually
	 * running as. They differ only while a task is temporarily assuming
	 * another's credentials, which is itself worth reporting accurately. */
	const struct cred *old = BPF_CORE_READ(task, real_cred);
	if (!old)
		return 0;

	__u32 old_uid = BPF_CORE_READ(old, uid.val);
	__u32 old_euid = BPF_CORE_READ(old, euid.val);
	__u64 old_caps = effective_caps(old);

	__u32 new_uid = BPF_CORE_READ(new, uid.val);
	__u32 new_euid = BPF_CORE_READ(new, euid.val);
	__u64 new_caps = effective_caps(new);

	__u32 flags = 0;
	if (new_euid == 0 && old_euid != 0)
		flags |= CRED_GAINED_ROOT;
	if (old_euid == 0 && new_euid != 0)
		flags |= CRED_LOST_ROOT;
	/* Gained, not merely changed: a task dropping capabilities is the
	 * hardening pattern every well-behaved daemon follows at startup, and
	 * reporting it would bury the one case that matters. */
	if (new_caps & ~old_caps)
		flags |= CRED_GAINED_CAPS;

	/* Nothing was gained. Credential changes that only drop privilege are the
	 * common case by a wide margin and are dropped here, before the cgroup
	 * lookup, so the hot path stays short. */
	if (!(flags & (CRED_GAINED_ROOT | CRED_GAINED_CAPS)))
		return 0;

	if (in_execve(task))
		flags |= CRED_IN_EXECVE;

	__u64 cgroup_id = bpf_get_current_cgroup_id();

	/* Enforcement: kill the task before the new credentials are used.
	 *
	 * Only for governed cgroups, and by default only when no execve explains
	 * the change - killing every setuid binary in a container would break sudo,
	 * ping and passwd, which is not a security posture anybody asked for. */
	__u32 mode = MODE_OBSERVE;
	__u32 *gm = bpf_map_lookup_elem(&cred_governed, &cgroup_id);
	if (gm)
		mode = *gm;

	if (mode == MODE_KILL) {
		int execve_exempt = (flags & CRED_IN_EXECVE) && !cfg(2);
		if (!execve_exempt) {
			if (bpf_send_signal(9) == 0)
				flags |= CRED_KILLED;
		}
	}

	struct cred_event *e = bpf_ringbuf_reserve(&cred_events, sizeof(*e), 0);
	if (!e)
		return 0;

	__u64 pid_tgid = bpf_get_current_pid_tgid();
	e->cgroup_id = cgroup_id;
	e->timestamp_ns = bpf_ktime_get_ns();
	e->old_caps = old_caps;
	e->new_caps = new_caps;
	e->pid = pid_tgid >> 32;
	e->tid = (__u32)pid_tgid;
	e->old_uid = old_uid;
	e->new_uid = new_uid;
	e->old_euid = old_euid;
	e->new_euid = new_euid;
	e->flags = flags;
	bpf_get_current_comm(&e->comm, sizeof(e->comm));

	bpf_ringbuf_submit(e, 0);
	return 0;
}

File: pkg/ebpf/credmonitor.go
package ebpf

import (
	"encoding/binary"
	"errors"
	"fmt"
	"strings"

	"github.com/cilium/ebpf"
)

// Flags on a CredEvent, mirroring bpf/cred_monitor.c. The two halves must move
// together.
const (
	// CredGainedRoot means the effective uid became 0 and was not 0 before.
	CredGainedRoot uint32 = 0x01
	// CredGainedCaps means the effective capability set grew.
	CredGainedCaps uint32 = 0x02
	// CredInExecve means an execve was underway, so a setuid binary explains
	// the change. Its absence on an escalation is the interesting case.
	CredInExecve uint32 = 0x04
	// CredKilled means the task was sent SIGKILL by the kernel program.
	CredKilled uint32 = 0x08
	// CredLostRoot means the effective uid stopped being 0 - a daemon
	// dropping privilege, reported only when something was also gained.
	CredLostRoot uint32 = 0x10
)

// CredEvent is a credential change observed by the kprobe on commit_creds.
//
// Every other event in Pahlevan describes a request the kernel was asked to
// satisfy. This one describes a result: the moment a task's privileges
// actually changed, whatever route it took. A kernel exploit that overwrites a
// cred struct and calls commit_creds directly produces one of these and no
// syscall event at all.
type CredEvent struct {
	CgroupID    uint64
	Timestamp   uint64
	OldCaps     uint64
	NewCaps     uint64
	PID         uint32
	TID         uint32
	OldUID      uint32
	NewUID      uint32
	OldEUID     uint32
	NewEUID     uint32
	Flags       uint32
	Comm        string
	ContainerID string
}

// Denied reports whether the kernel acted on this event. Named to match the
// other event types, where the same question is asked of a Flags field.
func (e *CredEvent) Denied() bool { return e.Flags&CredKilled != 0 }

// Unexplained reports whether privilege was gained with no execve underway.
//
// This is the discriminator the whole monitor exists for. A setuid binary
// gains privilege inside execve, and sudo, ping and passwd all do it dozens of
// times a day in an ordinary container. A task whose credentials change with
// no execve in progress either called a setuid syscall - which the syscall
// monitor reports, with its arguments - or did something the kernel has no
// other name for.
func (e *CredEvent) Unexplained() bool {
	return e.Flags&(CredGainedRoot|CredGainedCaps) != 0 && e.Flags&CredInExecve == 0
}

// GainedCapabilities lists the capabilities added by this change, by name.
func (e *CredEvent) GainedCapabilities() []string {
	return CapabilityNames(e.NewCaps &^ e.OldCaps)
}

// Summary renders the event the way an operator reads it.
func (e *CredEvent) Summary() string {
	var b strings.Builder
	fmt.Fprintf(&b, "%s[%d] uid %d->%d euid %d->%d", e.Comm, e.PID, e.OldUID, e.NewUID, e.OldEUID, e.NewEUID)
	if caps := e.GainedCapabilities(); len(caps) > 0 {
		fmt.Fprintf(&b, " gained %s", strings.Join(caps, ","))
	}
	if e.Flags&CredInExecve != 0 {
		b.WriteString(" (during execve)")
	} else {
		b.WriteString(" (no execve underway)")
	}
	if e.Flags&CredKilled != 0 {
		b.WriteString(" KILLED")
	}
	return b.String()
}

// CredEventHandler receives credential-change events.
//
// Separate from EventHandler and satisfied by type assertion, so a handler
// that has no interest in credential changes is not made to stub a method for
// them. Every registered handler is offered the event; those that do not
// implement this are skipped.
type CredEventHandler interface {
	HandleCredEvent(event *CredEvent) error
}

// parseCredEvent decodes `struct cred_event` from bpf/cred_monitor.c:
//
//	__u64 cgroup_id; __u64 timestamp_ns; __u64 old_caps; __u64 new_caps;
//	__u32 pid; __u32 tid; __u32 old_uid; __u32 new_uid;
//	__u32 old_euid; __u32 new_euid; __u32 flags; __u8 comm[16];  // 60 -> 64
func parseCredEvent(data []byte) *CredEvent {
	const commOff = 8 + 8 + 8 + 8 + 4*7 // 60
	const size = commOff + 16           // 76
	if len(data) < size {
		return nil
	}
	u32 := func(off int) uint32 { return binary.LittleEndian.Uint32(data[off : off+4]) }

	e := &CredEvent{
		CgroupID:  binary.LittleEndian.Uint64(data[0:8]),
		Timestamp: binary.LittleEndian.Uint64(data[8:16]),
		OldCaps:   binary.LittleEndian.Uint64(data[16:24]),
		NewCaps:   binary.LittleEndian.Uint64(data[24:32]),
		PID:       u32(32),
		TID:       u32(36),
		OldUID:    u32(40),
		NewUID:    u32(44),
		OldEUID:   u32(48),
		NewEUID:   u32(52),
		Flags:     u32(56),
	}
	comm := data[commOff : commOff+16]
	if i := indexZero(comm); i >= 0 {
		comm = comm[:i]
	}
	e.Comm = string(comm)
	e.ContainerID = fmt.Sprintf("cgroup:%d", e.CgroupID)
	return e
}

// setCredEnabled turns the kprobe's body on or off without detaching it.
//
// Detaching and reattaching a kprobe is not free and briefly leaves a window
// with no coverage; a config flag the program reads first costs one array
// lookup and has no window.
func (m *Manager) setCredEnabled(on bool) error {
	if m.credCollection == nil {
		return fmt.Errorf("the credential monitor is not loaded")
	}
	mp := m.credCollection.Maps["cred_config"]
	if mp == nil {
		return fmt.Errorf("the cred_config map is absent from the loaded object")
	}
	var key uint32
	var val uint32
	if on {
		val = 1
	}
	return mp.Put(&key, &val)
}

// CredMode is what the kernel does when a governed cgroup gains privilege.
type CredMode uint32

const (
	// CredObserve reports the change and lets it stand.
	CredObserve CredMode = 0
	// CredKill sends SIGKILL to the task before the new credentials are used
	// for anything.
	//
	// This is the only enforcement available at this site: commit_creds
	// returns void and is called past the point of no return, so the change
	// itself cannot be refused. There is no LSM hook here to refuse it at
	// either, which is precisely why the kprobe is worth having.
	CredKill CredMode = 1
)

// GovernCredentials sets what happens when a cgroup's tasks gain privilege.
//
// Ungoverned cgroups are observed and never killed, which is the right default
// for a control whose only response is a signal: a false positive here kills a
// process rather than denying one operation. A cgroup is governed only once an
// operator has decided that this workload never legitimately escalates.
func (m *Manager) GovernCredentials(cgroupID uint64, mode CredMode) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.credCollection == nil {
		return fmt.Errorf("the credential monitor is not loaded")
	}
	mp := m.credCollection.Maps["cred_governed"]
	if mp == nil {
		return fmt.Errorf("the cred_governed map is absent from the loaded object")
	}
	key := cgroupID
	val := uint32(mode)
	return mp.Put(&key, &val)
}

// UngovernCredentials returns a cgroup to observation only.
func (m *Manager) UngovernCredentials(cgroupID uint64) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.credCollection == nil {
		return fmt.Errorf("the credential monitor is not loaded")
	}
	mp := m.credCollection.Maps["cred_governed"]
	if mp == nil {
		return fmt.Errorf("the cred_governed map is absent from the loaded object")
	}
	key := cgroupID
	// A cgroup that was never governed is already ungoverned; that is the
	// requested state, not a failure.
	if err := mp.Delete(&key); err != nil && !errors.Is(err, ebpf.ErrKeyNotExist) {
		return err
	}
	return nil
}

// SetCredKillDuringExecve controls whether a governed cgroup's setuid binaries
// are killed too.
//
// Off by default and it should stay off almost everywhere: killing every
// setuid binary in a container breaks sudo, ping and passwd, which is not a
// security posture anybody asked for. It is worth having for a container that
// is known to contain no legitimate setuid binary at all, where any execve
// that gains privilege is by definition unexpected.
func (m *Manager) SetCredKillDuringExecve(on bool) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.credCollection == nil {
		return fmt.Errorf("the credential monitor is not loaded")
	}
	mp := m.credCollection.Maps["cred_config"]
	if mp == nil {
		return fmt.Errorf("the cred_config map is absent from the loaded object")
	}
	key := uint32(2)
	var val uint32
	if on {
		val = 1
	}
	return mp.Put(&key, &val)
}
